### DISCUSSION TITLE
Thread safety for tagging

### DISCUSSION THREAD

**Discussion by krystianzybala** [getsentry/sentry-java#2409] [Q&A] [answered]
Hi, 

quick question I have a multithread application that we use to Sentry to capture our exception. 
We have the REST app so I have a question about the thread safety of ```Sentry.setTag();``` before capturing the exception is it thread safety? 

```java 


class BusinessLogic {
    public void doSomething(final UserRegion region) {
        try {
            doBusinessLogic();
        } catch(Exception exception) {
            Sentry.setTag("user_region", region.current());
            Sentry.captureException(exception);
        }
    }

}
```

I mean a situation where two different threads with different UserRegion call that method but ```Sentry.setTag``` is global. In that case, the tag will be correctly propagated to the proper exception? 
Maybe a better approach is capturing the event and passing that information to a particular event.

**Comment by adinauer** [comment, marked as answer]
Hello @krystianzybala. Each thread has their own `Scope`. However the tags you set remain on the scope until removed or until the scope is replaced / popped (e.g. at the end of a request). This code should do what you are looking for:


```java
class BusinessLogic {
    public void doSomething(final UserRegion region) {
        try {
            doBusinessLogic();
        } catch(Exception exception) {
      		Sentry.withScope(scope -> {
        		scope.setTag("user_region", region.current());
            	Sentry.captureException(exception);
      		});
        }
    }
}
```

This only sets the tag for one call by creating a scope, setting the tag, capturing the exception and then popping the scope.

Here's some more docs on the topic: https://docs.sentry.io/platforms/java/enriching-events/scopes/

**Comment by krystianzybala** [comment]
Thank you :)
